File: ios/flutter_compress/Sources/flutter_compress/CompressionConfig.swift
import Foundation

/// Parsed mirror of the Dart `VideoCompressConfig`.
struct CompressionConfig {
  let quality: String
  let qualityPercent: Int?
  let targetSizeMB: Int?
  let videoBitrateKbps: Int?
  let codec: String
  let maxWidth: Int?
  let maxHeight: Int?
  let frameRate: Double?
  let removeAudio: Bool
  let audioBitrateKbps: Int?
  let trimStartMs: Int64?
  let trimEndMs: Int64?
  let alignment: String
  let keepOriginalIfLarger: Bool
  let minSavingsPercent: Int
  /// "auto" (keep source container where possible) or "mp4".
  let container: String

  init(map: [String: Any]) {
    quality = map["quality"] as? String ?? "medium"
    qualityPercent = (map["qualityPercent"] as? NSNumber)?.intValue
    targetSizeMB = (map["targetSizeMB"] as? NSNumber)?.intValue
    videoBitrateKbps = (map["videoBitrateKbps"] as? NSNumber)?.intValue
    codec = map["codec"] as? String ?? "h265"
    maxWidth = (map["maxWidth"] as? NSNumber)?.intValue
    maxHeight = (map["maxHeight"] as? NSNumber)?.intValue
    frameRate = (map["frameRate"] as? NSNumber)?.doubleValue
    removeAudio = map["removeAudio"] as? Bool ?? false
    audioBitrateKbps = (map["audioBitrateKbps"] as? NSNumber)?.intValue
    let trim = map["trim"] as? [String: Any]
    trimStartMs = (trim?["startMs"] as? NSNumber)?.int64Value
    trimEndMs = (trim?["endMs"] as? NSNumber)?.int64Value
    alignment = map["alignment"] as? String ?? "auto16"
    keepOriginalIfLarger = map["keepOriginalIfLarger"] as? Bool ?? true
    minSavingsPercent = (map["minSavingsPercent"] as? NSNumber)?.intValue ?? 0
    container = map["container"] as? String ?? "auto"
  }
}

/// The three ways a `compress` call can end.
enum CompressionOutcome {
  case success([String: Any])
  case cancelled
  case failure(String)
}

File: android/src/main/kotlin/com/compress/all/flutter_compress_pro/CompressionConfig.kt
package com.compress.all.flutter_compress_pro

/** Parsed mirror of the Dart [VideoCompressConfig]. */
data class CompressionConfig(
    val quality: String,
    val qualityPercent: Int?,
    val targetSizeMB: Int?,
    val videoBitrateKbps: Int?,
    val codec: String,
    val maxWidth: Int?,
    val maxHeight: Int?,
    val frameRate: Double?,
    val removeAudio: Boolean,
    val audioBitrateKbps: Int?,
    val trimStartMs: Long?,
    val trimEndMs: Long?,
    val alignment: String,
    val keepOriginalIfLarger: Boolean,
    val minSavingsPercent: Int,
    val notification: CompressionService.NotificationSpec?,
) {
    companion object {
        fun fromMap(m: Map<String, Any?>): CompressionConfig {
            @Suppress("UNCHECKED_CAST")
            val trim = m["trim"] as? Map<String, Any?>

            @Suppress("UNCHECKED_CAST")
            val notification = m["androidNotification"] as? Map<String, Any?>
            return CompressionConfig(
                quality = m["quality"] as? String ?: "medium",
                qualityPercent = (m["qualityPercent"] as? Number)?.toInt(),
                targetSizeMB = (m["targetSizeMB"] as? Number)?.toInt(),
                videoBitrateKbps = (m["videoBitrateKbps"] as? Number)?.toInt(),
                codec = m["codec"] as? String ?: "h265",
                maxWidth = (m["maxWidth"] as? Number)?.toInt(),
                maxHeight = (m["maxHeight"] as? Number)?.toInt(),
                frameRate = (m["frameRate"] as? Number)?.toDouble(),
                removeAudio = m["removeAudio"] as? Boolean ?: false,
                audioBitrateKbps = (m["audioBitrateKbps"] as? Number)?.toInt(),
                trimStartMs = (trim?.get("startMs") as? Number)?.toLong(),
                trimEndMs = (trim?.get("endMs") as? Number)?.toLong(),
                alignment = m["alignment"] as? String ?: "auto16",
                keepOriginalIfLarger = m["keepOriginalIfLarger"] as? Boolean ?: true,
                minSavingsPercent = (m["minSavingsPercent"] as? Number)?.toInt() ?: 0,
                notification = CompressionService.NotificationSpec.fromMap(notification),
            )
        }
    }
}

/**
 * The host app did not declare a permission the operation needs. The plugin
 * declares none of its own, so this always means "the app has to opt in".
 */
internal class PermissionDeniedException(message: String) : SecurityException(message)

/** Thrown when a job is cancelled. */
class CompressionCancelledException(message: String = "Compression cancelled") :
    Exception(message)
